MDANSE/Src/MDANSE/Framework/Jobs/TrajectoryFilter.py

```diff
@@ -0,0 +1,232 @@
+
+        self.numberOfSteps = self.configuration["atom_selection"]["selection_length"]
+
+        self._atoms = sorted_atoms(
+            self.configuration["trajectory"]["instance"].chemical_system.atom_list
+        )
+        self._selected_atoms = []
+        for indexes in self.configuration["atom_selection"]["indexes"]:
+            for idx in indexes:
+                self._selected_atoms.append(self._atoms[idx])
+        self._selected_atoms = AtomGroup(self._selected_atoms)
+
+        self.atomic_trajectory_array = np.zeros_like(
+            np.zeros(3*len(self._atoms)*len(self.configuration["frames"]["value"])).reshape((
+                len(self._atoms),
+                3,
+                len(self.configuration["frames"]["value"])
+            ))
+        )
+
+    def run_step(self, index):
+        """
+        Runs a single step of the job.\n
+
+        :Parameters:
+            #. index (int): The index of the step.
+        """
+        LOG.debug(f"Running step: {index}")
+        trajectory = self.configuration["trajectory"]["instance"]
+
+        # get atom index
+        indexes = self.configuration["atom_selection"]["indexes"][index]
+        atoms = [self._atoms[idx] for idx in indexes]
+
+        series = trajectory.read_com_trajectory(
+            atoms,
+            first=self.configuration["frames"]["first"],
+            last=self.configuration["frames"]["last"] + 1,
+            step=self.configuration["frames"]["step"],
+        )
+
+        self.atomic_trajectory_array[index] = series.T
+        return index, None
+
+    def combine(self, index, x):
+        """
+        Combines returned results of run_step.\n
+        :Parameters:
+            #. index (int): The index of the step.\n
+            #. x (any): The returned result(s) of run_step
+        """
+        pass
+
+    def finalize(self):
+        """
+        Finalizes the calculations (e.g. averaging the total term, output files creations ...).
+        """
+
+        # Get filter class and instantiate filter object
+        filter_config = json.loads(self.configuration["trajectory_filter"]['value'])
+        filter_class, filter_attributes = filter_map[filter_config["filter"]], filter_config["attributes"]
+
+        if {"n_steps", "time_step_ps"} not in set(filter_attributes.keys()):
+            filter_attributes.update({
+                "n_steps": self.configuration["trajectory"]["length"],
+                "time_step_ps": self.configuration["trajectory"]["md_time_step"],
+            })
+
+        filter = filter_class(**filter_attributes)
+
+        trajectories = copy.deepcopy(self.atomic_trajectory_array)
+
+        # Apply filter
+        filtered_coords = apply(filter, trajectories)
+
+        # Create trajectory writer object
+        self._output_trajectory = TrajectoryWriter(
+            self.configuration["output_files"]["file"],
+            self.configuration["trajectory"]["instance"].chemical_system,
+            self.numberOfSteps,
```

Comment:
I think there is a mistake here, the input parameter n_steps should be the number of MD frames. self.numberOfSteps is the number of selected atoms in this job.

Reply:
Yeah, good catch thanks - I have implemented this just now, and the results are looking much more sensible